### ISSUE TITLE
🧠 Implement Neural Network Layers

### ISSUE THREAD

**Issue by sakshambedi** [sakshambedi/micrograd#18]
# 🧠 Implement Neural Network Layers

## 📋 Description

Implement fundamental neural network layer abstractions to enable building and training neural networks with micrograd. This includes linear layers, activation functions, and the foundation for more complex architectures.

## 🎯 Acceptance Criteria

- [ ] Implement `Linear` (Dense/Fully Connected) layer
- [ ] Create base `Module` class for all layers
- [ ] Add `Parameter` wrapper for trainable tensors
- [ ] Implement activation layers: `ReLU`, `Sigmoid`, `Tanh`
- [ ] Add `Sequential` container for chaining layers
- [ ] Implement proper parameter initialization methods
- [ ] Add `parameters()` method to collect all trainable parameters
- [ ] Support `train()` and `eval()` modes
- [ ] Add proper gradient flow through layers
- [ ] Implement `zero_grad()` for all parameters

### Architecture Design

**Base Module Class**
```python
class Module:
    """Base class for all neural network modules."""
    
    def __init__(self):
        self._parameters = {}
        self._modules = {}
        self.training = True
    
    def __call__(self, *args, **kwargs):
        return self.forward(*args, **kwargs)
    
    def forward(self, *args, **kwargs):
        raise NotImplementedError
    
    def parameters(self):
        """Yield all parameters in this module and submodules."""
        for param in self._parameters.values():
            yield param
        for module in self._modules.values():
            yield from module.parameters()
    
    def zero_grad(self):
        """Zero gradients for all parameters."""
        for param in self.parameters():
            param.zero_grad()
    
    def train(self, mode=True):
        """Set training mode."""
        self.training = mode
        for module in self._modules.values():
            module.train(mode)
    
    def eval(self):
        """Set evaluation mode."""
        self.train(False)
```

**Parameter Class**
```python
class Parameter(Tensor):
    """A tensor that's considered a module parameter."""
    
    def __init__(self, data, requires_grad=True):
        super().__init__(data, requires_grad=requires_grad)
    
    def __repr__(self):
        return f"Parameter({super().__repr__()})"
```

**Linear Layer**
```python
class Linear(Module):
    """Linear transformation: y = xW^T + b"""
    
    def __init__(self, in_features, out_features, bias=True):
        super().__init__()
        self.in_features = in_features
        self.out_features = out_features
        
        # Initialize parameters
        self.weight = Parameter(Tensor.randn(out_features, in_features))
        if bias:
            self.bias = Parameter(Tensor.zeros(out_features))
        else:
            self.bias = None
        
        self.reset_parameters()
    
    def reset_parameters(self):
        """Initialize parameters using Xavier/Glorot initialization."""
        import math
        stdv = 1. / math.sqrt(self.weight.shape[1])
        self.weight.data.uniform_(-stdv, stdv)
        if self.bias is not None:
            self.bias.data.uniform_(-stdv, stdv)
    
    def forward(self, input):
        output = input @ self.weight.T
        if self.bias is not None:
            output = output + self.bias
        return output
```

### Activation Layers

**ReLU Activation**
```python
class ReLU(Module):
    """Rectified Linear Unit activation."""
    
    def forward(self, input):
        return input.relu()

class Sigmoid(Module):
    """Sigmoid activation."""
    
    def forward(self, input):
        return input.sigmoid()

class Tanh(Module):
    """Hyperbolic tangent activation."""
    
    def forward(self, input):
        return input.tanh()
```

**Sequential Container**
```python
class Sequential(Module):
    """Sequential container for layers."""
    
    def __init__(self, *args):
        super().__init__()
        for idx, module in enumerate(args):
            self._modules[str(idx)] = module
    
    def forward(self, input):
        for module in self._modules.values():
            input = module(input)
        return input
```

## 🧪 Testing Requirements

- [ ] Test Linear layer forward pass with different input/output sizes
- [ ] Test parameter initialization and shapes
- [ ] Test gradient flow through linear layers
- [ ] Test activation function layers
- [ ] Test Sequential container with multiple layers
- [ ] Test parameter collection and zero_grad functionality
- [ ] Test train/eval mode switching
- [ ] Test bias vs no-bias configurations
- [ ] Integration tests with optimizers
- [ ] Memory leak tests for large networks

## 📁 Files to Create/Modify

- `grad/nn/__init__.py` - Neural network package
- `grad/nn/module.py` - Base Module class
- `grad/nn/parameter.py` - Parameter class
- `grad/nn/linear.py` - Linear layer implementation
- `grad/nn/activation.py` - Activation functions
- `grad/nn/container.py` - Sequential and other containers
- `grad/nn/init.py` - Parameter initialization utilities
- `tests/test_nn_layers.py` - Comprehensive layer tests
- `tests/test_nn_parameter.py` - Parameter functionality tests
- `examples/simple_network.py` - Usage examples

## 🔗 Dependencies

- Requires: Matrix multiplication (#04)
- Requires: Autograd system (#01, #03)
- Requires: Additional tensor operations (#06)
- Blocks: Optimizers (#10)
- Blocks: Loss functions (#09)
- Blocks: Complete neural network training (#11)

### Basic Usage
```python
from grad.nn import Linear, ReLU, Sequential

# Define a simple network
model = Sequential(
    Linear(784, 128),
    ReLU(),
    Linear(128, 64),
    ReLU(), 
    Linear(64, 10)
)

# Forward pass
x = Tensor.randn(32, 784)  # Batch of 32 samples
output = model(x)          # Shape: (32, 10)

# Collect parameters
params = list(model.parameters())
print(f"Number of parameters: {len(params)}")

# Zero gradients
model.zero_grad()
```

### Training Loop Integration
```python
# Example training step
model = Sequential(Linear(10, 5), ReLU(), Linear(5, 1))
criterion = MSELoss()
optimizer = SGD(model.parameters(), lr=0.01)

x = Tensor.randn(32, 10)
y = Tensor.randn(32, 1)

# Forward pass
output = model(x)
loss = criterion(output, y)

# Backward pass
model.zero_grad()
loss.backward()

# Update parameters
optimizer.step()
```

### Custom Layer Example
```python
class Dropout(Module):
    def __init__(self, p=0.5):
        super().__init__()
        self.p = p
    
    def forward(self, input):
        if self.training:
            # Apply dropout during training
            mask = Tensor.rand(*input.shape) > self.p
            return input * mask / (1 - self.p)
        else:
            return input
```

## 🚨 Edge Cases to Handle

- Zero-dimensional inputs to Linear layers
- Very large layer sizes (memory management)
- Gradient explosion/vanishing through deep networks
- Parameter sharing across layers
- In-place operations affecting gradients
- Thread safety for parameter updates
- Proper cleanup of computational graphs

## 🔧 Performance Considerations

- Efficient parameter storage and access
- Memory-efficient gradient computation
- Minimize temporary allocations during forward/backward
- Consider future batch normalization requirements
- Optimize for common layer sizes
- Support for different initialization schemes

## 🏷️ Labels

- `enhancement`
- `neural-networks`
- `layers`
- `high-priority`
- `core-functionality`

## 📈 Priority

**High** - Essential for building and training neural networks

## 🔗 Related Issues

- Depends on: Matrix multiplication (#04)
- Depends on: Autograd system (#01, #03)
- Depends on: Tensor operations (#06)
- Blocks: Optimizers (#10)
- Blocks: Loss functions (#09)
- Blocks: Complete training pipeline (#11)
